hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/CombinedFileRange.java

```diff
@@ -81,7 +81,11 @@ public boolean merge(long otherOffset, long otherEnd, FileRange other,
     long end = this.getOffset() + this.getLength();
     long newEnd = Math.max(end, otherEnd);
     if (otherOffset - end >= minSeek || newEnd - this.getOffset() > maxSize) {
-      return false;
+      // Guarantee that the ranges are not overlapped in order to avoid HADOOP-19098 violation later.
+      // The case `maxSize==0` is excluded due to the explicit test case, testMaxSizeZeroDisablesMerging.
+      if (otherOffset >= end || maxSize == 0) {
```

Comment:
This will result in merging even when 'mergedRange > maxSize'. Are we planning to remove maxSize check and merge anyway?